src/client/dfs/metrics.c

```diff
@@ -0,0 +1,174 @@
+	}                                                                                          \
+	i++;
+
+static void
+op_stats_init(struct dfs_metrics *metrics, uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	char tmp_path[D_TM_MAX_NAME_LEN] = {0};
+	int  i                           = 0;
+	int  rc;
+
+	if (metrics == NULL)
+		return;
+
+	D_FOREACH_DFS_OP_STAT(ADD_STAT_METRIC);
+}
+
+static void
+cont_stats_init(struct dfs_metrics *metrics, uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	char tmp_path[D_TM_MAX_NAME_LEN] = {0};
+	int  rc                          = 0;
+
+	if (metrics == NULL)
+		return;
+
+	SPRINTF_TM_PATH(tmp_path, pool_uuid, cont_uuid, "mount_time");
+	rc = d_tm_add_metric(&metrics->dm_mount_time, D_TM_TIMESTAMP, "container mount time", NULL,
+			     tmp_path);
+	if (rc != 0)
+		DL_ERROR(rc, "failed to create mount_time timestamp");
+}
+
+static void
+file_stats_init(struct dfs_metrics *metrics, uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	char tmp_path[D_TM_MAX_NAME_LEN] = {0};
+	int  rc                          = 0;
+
+	if (metrics == NULL)
+		return;
+
+	SPRINTF_TM_PATH(tmp_path, pool_uuid, cont_uuid, DFS_METRICS_ROOT "/read_bytes");
+	rc = d_tm_add_metric(&metrics->dm_read_bytes, D_TM_STATS_GAUGE, "dfs read bytes", "bytes",
+			     tmp_path);
+	if (rc != 0)
+		DL_ERROR(rc, "failed to create dfs read_bytes counter");
+	rc =
+	    d_tm_init_histogram(metrics->dm_read_bytes, tmp_path, NR_SIZE_BUCKETS, 256, 2, "bytes");
+	if (rc)
+		DL_ERROR(rc, "Failed to init dfs read size histogram");
+
+	SPRINTF_TM_PATH(tmp_path, pool_uuid, cont_uuid, DFS_METRICS_ROOT "/write_bytes");
+	rc = d_tm_add_metric(&metrics->dm_write_bytes, D_TM_STATS_GAUGE, "dfs write bytes", "bytes",
+			     tmp_path);
+	if (rc != 0)
+		DL_ERROR(rc, "failed to create dfs write_bytes counter");
+	rc = d_tm_init_histogram(metrics->dm_write_bytes, tmp_path, NR_SIZE_BUCKETS, 256, 2,
+				 "bytes");
+	if (rc)
+		DL_ERROR(rc, "Failed to init dfs write size histogram");
+}
+
+bool
+dfs_metrics_enabled()
+{
+	/* set in client/api/metrics.c */
+	return daos_client_metric;
+}
+
+void
+dfs_metrics_init(dfs_t *dfs)
+{
+	uuid_t pool_uuid;
+	uuid_t cont_uuid;
+	char   root_name[D_TM_MAX_NAME_LEN];
+	pid_t  pid       = getpid();
+	size_t root_size = DFS_METRICS_SIZE + (D_TM_METRIC_SIZE * 3);
+	int    rc;
+
+	if (dfs == NULL)
+		return;
+
+	rc = dc_pool_hdl2uuid(dfs->poh, NULL, &pool_uuid);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to get pool UUID");
+		goto error;
+	}
+
+	rc = dc_cont_hdl2uuid(dfs->coh, NULL, &cont_uuid);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to get container UUID");
+		goto error;
+	}
+
+	snprintf(root_name, sizeof(root_name), "%d", pid);
+	/* if only container-level metrics are enabled; this will init a root for them */
+	rc = d_tm_init_with_name(d_tm_cli_pid_key(pid), root_size, D_TM_OPEN_OR_CREATE, root_name);
+	if (rc != 0 && rc != -DER_ALREADY) {
+		DL_ERROR(rc, "failed to init DFS metrics");
+		goto error;
+	}
```

Comment:
will all threads share the same root here? I wonder if all there are a lot dfuse service threads, will it reach to the root_size maximum? or it will create a share memory for each pid?

Reply:
They all share the same segment. I think this is OK, since we don't care about tracking per-thread metrics.